docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1020,6 +1021,173 @@ table my_ipv4_lkup {
+unstructured `annotations`. The `structured_annotations` contain parsed
+representations of the original annotation source. This parsing includes
+expression-evaluation, so the resulting `P4Info` may contain a simplified
+replica of the original structured annotations.
+
+The structured annotation messages are defined in `p4types.proto`.
+
+~ Begin Proto
+message KeyValuePair {
+  string key = 1;
+  Expression value = 2;
+}
+
+message KeyValuePairList {
+  repeated KeyValuePair kv_pairs = 1;
+}
+
+message Expression {
+  oneof value {
+    string string_value = 1;
+    int64 int64_value = 2;
+    bool bool_value = 3;
+  }
+}
+
+message ExpressionList {
+  repeated Expression expressions = 1;
+}
+
+message StructuredAnnotation {
+  string name = 1;
+  oneof body {
+    ExpressionList expression_list = 2;
+    KeyValuePairList kvpair_list = 3;
+  }
+}
+~ End Proto
+
+The `StructuredAnnotation` message can represent either a `KeyValuePairList`
+or an `ExpressionList`.
+
+The type of an expression is intentionally limited to one of the following
+base types: string literal, 64-bit signed integer, or boolean. The `p4c`
+compiler frontend which generates `P4Info` will evaluate all expressions and
+simplify them to one of the valid types. Any expressions which don't match one
+of the valid types will generate an error. For integers exceeding 64 bits,
+besides issuing an error, the compiler *could* print a suggestion to use a
+string representation, and the `P4Info` consumer could perform any necessary
+conversions.
+
+The following invariants hold:
+
+1. For any P4 entity, there are no two `StructuredAnnotation`s that have the
+same name.
+
+2. Within a `KeyValuePairList`, there are no two `KeyValuePair`s that have the
+same `key.`
+
+### Structured Annotation Examples
+
+**Empty Expression List**
+
+~ Begin P4Example
+@Empty[]
+table t {
+    ...
+}
+~ End P4Example
+
+The generated p4Info will contain:
+
+~ Begin Proto
+structured_annotations {
+  name: "Empty"
+}
+~ End Proto
+
+**Mixed Expression List**
+
+~ Begin P4Example
+#define TEXT_CONST "hello"
+#define NUM_CONST 6
+@MixedExprList[1,TEXT_CONST,true,1==2,5+NUM_CONST]
+table t {
+    ...
+}
+~ End P4Example
+
+The generated p4Info will contain:
+
+~ Begin Proto
+structured_annotations {
+  name: "MixedExprList"
+  expression_list {
+    expressions {
+      int64_value: 1
+    }
+    expressions {
+      string_value: "hello"
+    }
+    expressions {
+      bool_value: true
+    }
+    expressions {
+      bool_value: false
+    }
+    expressions {
+      int64_value: 11
+    }
+  }
+}
+~ End Proto
+~ End Proto
+
+**kvList of Mixed Expressions**
+
+~ Begin P4Example
+@MixedKV[label="text", my_bool=true, int_val=2*3]
+table t {
+    ...
+}
+~ End P4Example
+
+The generated p4Info will contain:
+
+~ Begin Proto
+structured_annotations {
+  name: "MixedKV"
+  kvpair_list {
+    kv_pairs {
+      key: "label"
+      value {
+        string_value: "text"
```

Comment:
These should be wrapped in a "value" message, I believe. One for kvPair and one for Expression.

Reply:
@stefanheule I don't understand this comment. I generated this output using Python generated _pb2.py libraries so I believe the structure is correct. Can you show ,e what you expected?

Reply:
Sorry, my bad, it is correct.